strawberryfields/tdm/tdmprogram.py

```diff
@@ -303,6 +304,91 @@ def context(self, *args, copies=1, shift="default"):
         self.loop_vars = self.params(*[f"p{i}" for i in range(len(args))])
         return self
 
+    def compile(self, *, device=None, compiler=None, **kwargs):
+        """Compile the program given a Strawberry Fields photonic hardware device specification.
+        At this stage the compilation is simply a check that the program matches the device.
+
+        Args:
+            device (~strawberryfields.api.DeviceSpec): device specification object to use for
+                program compilation
+            compiler (str, ~strawberryfields.compilers.Compiler): Compiler name or compile strategy
+                to use. If a device is specified, this overrides the compile strategy specified by
+                the hardware :class:`~.DevicSpec`.
+
+        Keyword Args:
+            optimize (bool): If True, try to optimize the program by merging and canceling gates.
+                The default is False.
+            warn_connected (bool): If True, the user is warned if the quantum circuit is not weakly
+                connected. The default is True.
+
+        Returns:
+            Program: compiled program
+        """
+        device_layout = bb.loads(device.layout)
+        # First check: the gates are in the correct order
+        program_gates = [cmd.op.__class__.__name__ for cmd in self.rolled_circuit]
+        device_gates = [op["op"] for op in device_layout.operations]
+        if device_gates != program_gates:
+            raise CircuitError(
+                "Program cannot be used with the device '{}' "
```

Comment:
Also happy to get suggestions in terms of more meaningful error messages

Reply:
Maybe something like `Program does not match the physical architecture of the device`? As per the wording from here https://strawberryfields.ai/photonics/demos/tutorial_X8.html#hardware-compilation

Reply:
Just saw the next check: maybe worth referring here to the gates and their expected order in this message. Even if we manage to pass this first check, the topology/architecture can still be different if gates act on different modes as expected (second check).